src/main/java/service/RegionFetcher.java

```diff
@@ -1,101 +1,155 @@
-
-
-                    if(regionServers.containsKey(serverName)) {
-                        regionServers.get(serverName).add(regionName);
-                    } else {
-                        regionServers.put(serverName, new ArrayList<>());
-                        regionServers.get(serverName).add(regionName);
-                    }
-                    lastRow = getNextStartRow(cell.getRow());
-                }
-            }
+      } else {
+        break;
+      }
+    }
+  }
+
+  private void refreshRegions() throws IOException {
+    allRegions.clear();
+    allRegions.addAll(
+        this.admin.getRegions(tableName).stream().map(region -> region.getEncodedName()).collect(Collectors.toList()));
+
+    for (ServerName sn : this.admin.getRegionServers()) {
+      List<RegionInfo> regions = this.admin.getRegions(sn);
+      regions.stream().forEach(region -> {
+        if (allRegions.contains(region.getEncodedName())) {
+          regionFNHostnameMapping.putIfAbsent(region.getEncodedName(), new ArrayList<String>() {{
+            sn.getHostname();
+          }});
         }
-        scanner.close();
-        return regionServers;
+      });
     }
-
-    private byte[] getNextStartRow(byte[] row) {
-        byte[] startRow = new byte[row.length + 1];
-
-        System.arraycopy(row, 0, startRow, 0, row.length);
-        startRow[row.length] = (byte) 0;
-        return startRow;
+    log.trace("All regions: " + allRegions);
+  }
+
+  private Scan getScan(byte[] cf, byte[] column, Optional<PrefixFilter> filterOptional) {
+    Scan scan = new Scan();
+    scan.addColumn(cf, column);
+    if (filterOptional.isPresent()) {
+      scan.setFilter(filterOptional.get());
+    }
+    return scan;
+  }
+
+  public List<String> getNextBatchOfEncodedRegions(Set<String> inProgressRegions) throws IOException {
+    iteration += 1;
+    log.info("Fetching regions for this iteration: " + iteration + ", regionsCompactionInProgress: " + inProgressRegions
+        .size() + ", regionsCompacted: " + (processedRegions.size() - inProgressRegions.size()) + ", allRegions: "
+        + allRegions.size());
+    List<String> encodedRegions = new ArrayList<>();
+    refreshRegions();
+    refreshFavoredNodesMapping();
+
+    Set<String> serversForThisBatch = new HashSet<>();
+    for (String encodedRegion : inProgressRegions) {
+      if (regionFNHostnameMapping.containsKey(encodedRegion)) {
+        serversForThisBatch.addAll(regionFNHostnameMapping.get(encodedRegion));
+      }
     }
 
-    public Scan getScan() {
-        Scan scan = new Scan();
-        scan.addColumn(Bytes.toBytes("info"), Bytes.toBytes("server"));
-        scan.setFilter(new PrefixFilter(Bytes.toBytes(tableName + ",")));
-
-        if(lastRow != null)
-            scan.setStartRow(lastRow);
-        return scan;
+    for (String encodedRegion : allRegions) {
+      if (!inProgressRegions.contains(encodedRegion) && encodedRegions.size() < (
```

Comment:
What is the need for iterating over `allRegions` ? We are invalidating a region with `regionFNHostnameMapping.containsKey` check. Lets directly iterate over `regionFNHostnameMapping`.

Adding to this, I think, method `refreshRegions` is unnecessary, since `regionFNHostnameMapping` is being populated by reading meta-table only.
Can there be scenario, where the region is there but not present in meta-table ?

Reply:
`What is the need for iterating over allRegions` <-- This is if regions merge / split

`Can there be scenario, where the region is there but not present in meta-table ?` <-- Not every deployment uses concept of favored nodes.